client/src/main/scala/org/http4s/client/middleware/DigestClient.scala

```diff
@@ -0,0 +1,104 @@
+/*
+ * Copyright 2013-2020 http4s.org
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.http4s.client.middleware
+
+import java.util.Base64
+
+import cats.data.NonEmptyList
+import cats.effect.{Concurrent, Resource}
+import cats.implicits.{none, _}
+import org.http4s.Status.Unauthorized
+import org.http4s._
+import org.http4s.client.Client
+import org.http4s.headers.{Authorization, `WWW-Authenticate`}
+import org.http4s.middleware.authentication.DigestUtil
+import org.http4s.parser.HttpHeaderParser
+import org.log4s.getLogger
+
+import scala.util.Random
+
+/**
+  * Digest Middleware for digest auth
+  */
+object DigestClient {
+  private[this] val logger = getLogger
+
+  def apply[F[_]](
+      username: String,
+      password: String
+  )(client: Client[F])(implicit F: Concurrent[F]): Client[F] =
+    apply[F](username, password, "auth")(client)
+
+  def apply[F[_]](
+      username: String,
+      password: String,
+      qop: String
+  )(client: Client[F])(implicit F: Concurrent[F]): Client[F] = {
+    def digestAuthHeader(req: Request[F], challenge: Challenge): Header = {
+      val realm = challenge.realm
+      val method = req.method.name
+      val uri = req.uri.path.toString
+      val nc = "00000001"
```

Comment:
nc and cnonce are in lockstep if memory serves, if you always provide a new cnonce, then you dont need to increase nc.
This should be verified with the rfc.

Reply:
> The nc parameter stands for "nonce count".  The nc value is the
      hexadecimal count of the number of requests (including the current
      request) that the client has sent with the nonce value in this
      request.  For example, in the first request sent in response to a
      given nonce value, the client sends "nc=00000001".  The purpose of
      this parameter is to allow the server to detect request replays by
      maintaining its own copy of this count -- if the same nc value is
      seen twice, then the request is a replay.

This is challenging to do right.  We could maintain a `ConcurrentMap[String, Long]` for the nonce counter, but it would grow indefinitely unless we provide an eviction method.  We could maintain a single counter, but some servers get cranky if the nonce counter skips numbers.  We could parameterize a `String => F[Long]`, but we should have a default implementation.

Reply:
I propose we add the following algebra:

```scala
trait NonceCounter[F[_]] {
  /** Returns the next `nc` value for the given nonce */
  def nc(nonce: String): F[Long]
  /** Call when a conversation is over, to prevent unbounded growth. */
  def evict(nonce: String): F[Unit]
}
```

Then we can provide an implementation around that with a `Sync[F]` constraint around a `concurrent.Map`.  @ChristopherDavenport has something like this in mules.  That implementation would be the default.  People who want something more sophisticated could then easily adopt it to a more industrial cache like mules.  Or who want a more lax implementation could create a constant version.

@kuzkdny, does this make sense?  Would you like to try adding this?  I think we should at least get that algebra in place before merging, so we don't break binary compatibility.

Reply:
@rossabaker , thanks. this all makes sense, I'll take a look this days.

Reply:
I suppose there also could be an `evictAll: F[Unit]` on that algebra.  I don't feel strongly about that either way, but it's another decision to consider before we merge into 0.21.

Reply:
Just take a closer look at [rfc7616](https://tools.ietf.org/html/rfc7616) will do some more update in client middleware and DigestUtils(that one that is new and really just a copy of server DigestUtils) to be better compatible(and avoid any next pull request binary compatibility issue)

Reply:
Hi @rossabaker I've pushed updates related to NonceCounter, have covered this with tests, but really will be happy this to be good reviewed.
`class SimpleInMemoryStoreNonceCounter[F[_]](nonceTtl: FiniteDuration, cnonceF: () => String)(
    implicit
    T: Timer[F],
    C: Concurrent[F])
    extends NonceCounter[F] {
  private val cache = TrieMap.empty[String, Ref[F, Entry]]
  override def nc(nonce: String): F[(String, Long)] =
    (for {
      now <- T.clock.realTime(TimeUnit.MILLISECONDS)
      ref <- cache.getOrElseUpdate(nonce, Ref.unsafe(Entry(now, cnonceF(), 0))).pure[F]
      next <- ref.updateAndGet(d => d.copy(usedAt = now, nc = d.nc + 1))
      res <-
        if (next.nc > Max) ref.get >>= (_ => ref.updateAndGet(_ => Entry(now, cnonceF(), 1)))
        else next.pure[F]
    } yield (res.cnonce, res.nc)) <* C.start(autoCleanIfNotUsed(nonce))

  private def autoCleanIfNotUsed(nonce: String): F[Unit] =
    for {
      time <- T.clock.realTime(TimeUnit.MILLISECONDS)
      _ <- T.sleep(nonceTtl)
      _ <- cache.get(nonce) match {
        case Some(ref) => ref.get >>= (r => if (r.usedAt > time) ().pure[F] else evict(nonce))
        case None => ().pure[F]
      }
    } yield ()

  override def evict(nonce: String): F[Unit] =
    cache.remove(nonce).map(_ => ().pure[F]).getOrElse(().pure[F])

  override def evictAll: F[Unit] =
    cache.clear.pure[F]

  case class Entry(usedAt: Long, cnonce: String, nc: Long)
}`

 I'll try to find some time this week to make client to be fully compatible with rfc, and sure to be green line